### ISSUE TITLE
Auto dictionary lookup after long tap

### ISSUE THREAD

**Issue by rtega** [koreader/koreader#4091]
* KOReader version: v2015.11-1735
* Device: Kobo

#### Issue

When reading Japanese I have to select words by long tap because words are not separated in the text. I have to tap "Dictionary" time and again to do a lookup. Is there a way to have koreader lookup in the dictionary immediately?

**Comment by poire-z** [comment]
The current text selection behaviour is this:
- if one single word is selected: launch the dict lookup (as most dictionaries have only entries for single individual words, and that's what users most probably wish when holding on a single word)
- if more than one word is selected: show the little menu allowing to highlight, etc...

For western languages, that works really well.

Now, with Chinese, Japanese and Korean (CJK), there is something everywhere in the code that decides that __any such CJK char is itself a single word__ (and the HTML/CSS specs say the same).
I don't know real much about these languages, and if really multiple CJK chars/ideographs can make out a single word. Can they? In general or only because there are entries with multiple chars in some dictionaries? Or in all dictionaries?

If that's possible generally:
Are there some rules to detect word boundaries, like which CJK (or J for you :) char expect left or right neighbour char to be part of the same word?
I see we have a list of J chars that are not supposed to be splitted on a next line (used when laying out the text on lines, wrapping...):
https://github.com/koreader/koreader/blob/9971eb8533528cbb86707d81770ba26b1b8aaa9b/frontend/util.lua#L265-L267
(click on the link to util.lua and look around for other CJK chars that should not be cut on the left or on the right only).
Do the words you lookup include these chars? In which case, these could be the rules.

Otherwise, we could add an option to Dictionary settings to go do a dict lookup even if multiple words are selected (thus not easy to implement rightly, I had a try...)

**Comment by Frenzie** [comment]
> I don't know real much about these languages, and if really multiple CJK chars/ideographs can make out a single word. Can they? In general or only because there are entries with multiple chars in some dictionaries? Or in all dictionaries?

Characters can are basically syllables, and most words are polysyllabic. Besides, simple logic dictates a few hundred is probably the utmost limit of monosyllabic words, even with the Sinetic up/down padding.

**Comment by rtega** [comment]
There exists a program like mecab (https://en.wikipedia.org/wiki/MeCab) which can split up a japanese sentence in words. On linux the program is about 6Kb  but the dictionaries it uses are about 75Mb.

**Comment by poire-z** [comment]
> I have written a bunch of code to de-inflect Japanese verbs and adverbs to their original form

Dunno if that's the job of koreader or of the stardict lookup tool, or of the startdict dictionary itself.
Stardict dictionaries can have synonyms to help with this, and there's the fuzzy search that helps for european languages where often only the end of the word changes.
(It usually works well, but I've seen some cases where for some strange past tense form of a verb, I got some lookup results with anything but that verb :)

> some kind of framework where we could set a language-state 

That would be quite some work (discussed in #2476), but we can get the language from the document's metadata, which may be absent with some documents. You can see it in `Book information`, to see if it's there with your books.
`frontend/apps/reader/modules/readerhyphenation.lua` uses it to decide which hyphenation dict to use.

Another solution would be to guess the language from the word, by looking at the first chars - I guess that would be 100% accurate with detecting japanese words are japanese :)
We have a `ReaderDictionary:cleanSelection(text)` that does some clean-up. It could be made to do additional cleanup or normalisation, if you're sure what it does is fine for any dictionary (possibly with another dictionary setting to enable/disable this additional normalisation).

> a extra lua file like we do the make-up for certain dictionaries right now.

That happens **after** the lookup, result by result, and we know from which dict comes the result, and can use dict specific stuff.
Doing what you want **before** the lookup would be more complicated: for each dict that is enabled and will be looked up: see if it has that pre_lookup_func/file and fix the input word, and if the word has been changed, do a dedicated lookup - which will increase the lookup time (and complexity in readerdictionary.lua).

**Comment by rtega** [comment]
> > I have written a bunch of code to de-inflect Japanese verbs and adverbs to their original form
> Dunno if that's the job of koreader or of the stardict lookup tool, or of the startdict dictionary itself. Stardict dictionaries can have synonyms to help with this, and there's the fuzzy search that helps for european languages where often only the end of the word changes. (It usually works well, but I've seen some cases where for some strange past tense form of a verb, I got some lookup results with anything but that verb :)
Problem is that on my device lookups with synonyms takes ages. My only option right now is to straighten the dictionary out. With the amount of possibilities for a japanese verb this is out of the question. Or has the issue with synonyms in sdcv already been solved?

Regarding #2476 wouldn't it at least be handy to have dictionary groups? I have several dictionaries per language that I read. When switching between different books it takes quite a while to select the correct dictionaries.

**Comment by Frenzie** [comment]
Is it that rare? I read in four languages, although possibly not in
KOReader,[1] and I don't particularly need my French dictionaries
interfering with English. It's not really *bothersome* to me or I would've
acted on it long ago, but it doesn't feel like a niche requirement to me.

[1] I don't recall ever loading a German PDF onto my H2O.

**Comment by rtega** [comment]
Something like a pre-lookup would be nice. I'll attach the readerdictionary.lua-file as I've implemented it for Japanese now (apparently I can't attach lua-files).
[readerdictionary.lua.zip](https://github.com/koreader/koreader/files/2369616/readerdictionary.lua.zip)

One thing would be at least very helpful: keep dictionary selections saved per document, just like font settings and the like. I regularly switch between books in different languages and having to reselect all the dictionaries is quite cumbersome each.

**Comment by cyphar** [comment]
Did the pre-lookup hook mentioned in [here](https://github.com/koreader/koreader/issues/4091#issuecomment-417635334) ever get implemented? Looking at the code, it doesn't seem like it. I'd be interested in contributing something like that, because the ability to effectively use KOReader for Japanese word lookups kinda depends on there being some way to get proper deinflection (because fuzzy matching causes lots of issues with Japanese lookups and while there are tricks you can do to). I ran into some issues with text selection #8270 which are made worse by the fact that there is no plugin for Japanese verb deinflection and no real way to write one at the moment.

I think it would be in theory great if koreader could use mecab to make text selection nicer (right now Japanese readers are stuck having to manually select words to look up and the tap-to-lookup functionality is actually more of a hindrance than a help because you have to close the dictionary pop-up to try to select the text again), but I can understand not wanting to do it.

**Comment by rtega** [comment]
Not as far as I know.

**Comment by Frenzie** [comment]
If I understand correctly all it needs is iconv? https://taku910.github.io/mecab/#install-unix It's a bit unclear to me where to find the actual sources to try to gauge impact though.

**Comment by poire-z** [comment]
> Did the pre-lookup hook mentioned in here ever get implemented? Looking at the code, it doesn't seem like it.

No, it hasn't. Again, we're not familiar with these languages and the issues, we can just wait for someone to come up with at least an idea of the API and the start of some pre_lookup.lua implementation. Then, we see what input is needed, what output we can get - and try to think a bit more generically so this hook could be used for other ideas - and then, the KOReader "core" part shouldn't be much and should be easy to add.

(https://github.com/koreader/koreader/issues/4091#issuecomment-420164656 felt not really like that generic idea - may be it could be shoved into another module, but frankly, I don't have the mind opened enough to think about how to generically extend this so it would work not only for japanese... dunno if @rtega  has continued it or is using it daily.)

**Comment by rtega** [comment]
> ([#4091 (comment)](https://github.com/koreader/koreader/issues/4091#issuecomment-420164656) felt not really like that generic idea - may be it could be shoved into another module, but frankly, I don't have the mind opened enough to think about how to generically extend this so it would work not only for japanese... dunno if @rtega has continued it or is using it daily.)

I haven't found the time to look into this. At the moment I'm not reading any Japanese either though the code would still be valid.

**Comment by cyphar** [comment]
I don't know about about every synthetic/agglutinative language there is, but I think that having a few hooks which could be used to provide language-specific handling for selection would be incredibly useful and should be generic enough for a first draft. The most generic (and simple) idea I came up with is the following set of hooks for a hypothetical Japanese plugin:

 1. On single tap, have a hook which any language-specific plugin would be able to use in order to specify where the word boundary most likely is (the most straightforward way of doing this would probably be to give the hook all of the text on the current page starting from the tapped character -- or possibly all of the text up to the next punctuation mark). With Japanese this would be enough for you to run mecab to do text segmentation and that would give you the correct word boundary in most cases. The reason this would ideally be run when you do the tap rather than pre-computed per-page is simply that because mecab can make mistakes in text segmentation, a mistake earlier in the page would cause later words to be incorrectly segmented.
 2. On a long tap to create a selection, the selection should default to the short-tap selection by default but then should be controllable so that you could adjust the lookup if mecab incorrectly segmented the text.
 3. However, in order for these two features to be usable for Japanese word lookups (and likely any agglutinative language), it is necessary to have the `pre_lookup.lua` hook mentioned above because as I mentioned above, StarDict's fuzzy searching is not useful for searching inflected Japanese verbs because it ends up throwing away the actual verb. The lookup would basically just need to take the selection text and it'd return the dictionary form to look up in the dictionary (this has the added benefit that if you selected too many characters, the deinflection hook could actually clean up the lookup).
 4. In order for the plugin to work smoothly, we probably need to have a way of specifying the document language to pick the correct language-specific plugin, which means that we'd either need to try to auto-detect the language, or ask the user to specify the language.

I don't know how practical this would be with PDFs (my layman understanding is that it is incredibly difficult to do accurate text extraction from an arbitrary PDF document so I don't know how koreader currently handles the problem, and whether it's possible to have the above hooking system work with PDFs). Personally I only read EPUBs so I'd need to look a bit more into it.

> Again, we're not familiar with these languages and the issues, we can just wait for someone to come up with at least an idea of the API and the start of some pre_lookup.lua implementation.

That's completely understandable -- and to be clear, since this is something I would like to have, I would be interested in implementing it since I fully appreciate it's very difficult to write language-specific features for a language you know nothing about -- I'm just sketching out ideas to see whether this is something you'd all find reasonable.

Some follow-up questions: Given folks learning other languages might be interested in this -- how strong are your API guarantees for hooks? Would the Japanese plugin live in the koreader source or would it have to mantained separately? If it's maintained separately, is there some documentation on how to maintain a plugin which needs binaries alongside it (looking at the koreader source it's clear that koreader executes external programs such as `sdcv` internally but I have no idea how they're packaged and built for the different platforms koreader runs on)?

**Comment by Frenzie** [comment]
> Some follow-up questions: Given folks learning other languages might be interested in this -- how strong are your API guarantees for hooks? 

We put a fair bit of effort into not breaking things.

> Would the Japanese plugin live in the koreader source or would it have to mantained separately?

This belongs in the main repo, doesn't it?

>(looking at the koreader source it's clear that koreader executes external programs such as sdcv internally but I have no idea how they're packaged and built for the different platforms koreader runs on)?

See https://github.com/koreader/koreader-base, specifically the Makefiles and /thirdparty. But for basic development all you need is the library. If there's any difficulty, it's most likely to come from Android.

> If it's too difficult to package

What's difficult is that it's all Japanese. ;-) It looks like the mecab library compiles to about 2 MB, which is a bit on the larger size relatively speaking but I don't think that should be a problem. The 50+ MB of data would have to be downloaded separately though.

Is it a problem that it says the last update was from 2013?

**Comment by poire-z** [comment]
Well, I'm not sure I'd want this japanese thingy (lib+data files) into the main repo.... and japanese grammar discussions in our main project :)
For a mecab build, we might not have a choice if we want it compiled for all the platforms but include it in koreader-base - but we're not there yet :) We'd need first to get an idea on what/how it could integrate.

Looking around, there might be alternative ? Or at least, stuff that seem to apply to other languages ?
`https://github.com/kaegi/MorphMan/issues/225`
I have little interest in this languages grammar topics, but if we do something, we should aim at having generic hooks so it works not only with Japanese.

What's @rtega effort in https://github.com/koreader/koreader/issues/4091#issuecomment-420164656 all about ? How does it compare to using Mecab ? Is it a Mecab-light ? Could a small pure Lua solution like this be enough ? (a yes/no answer will suffice :)
Or could Mecab be rewritten in Lua ? 

(Sorry, not much brain power available currently to digest the above posts and answer longer :)

**Comment by cyphar** [comment]
@poire-z 

> (Sorry, not much brain power available currently to digest the above posts and answer longer :)

Sorry, I'll try to be more concise. :sweat_smile: 

> Well, I'm not sure I'd want this japanese thingy (lib+data files) into the main repo.... and japanese grammar discussions in our main project :) For a mecab build, we might not have a choice if we want it compiled for all the platforms but include it in koreader-base - but we're not there yet :) We'd need first to get an idea on what/how it could integrate.

Completely understandable. I would mind maintaining the Japanese-specific portions of this outside of the repo if that is preferable. With the right set of hooks it might not be necessary at all for koreader itself to care about Japanese at all.

The main question is would I be able to externally package MeCab with an out-of-tree extension or if it is necessary for this to be in `koreader-base`. If it is possible for an external plugin to package binaries properly then koreader itself doesn't need to care about these implementation details and only the hooks are necessary.

> I have little interest in this languages grammar topics, but if we do something, we should aim at having generic hooks so it works not only with Japanese.

Agreed, I tried to make my above proposal generic enough it should work for other languages. I'll ping some Korean learners I know to ask their opinion. I am trying to answer your questions in a way that mentions Japanese grammar as little as possible, sorry if it's still too much. :bowing_man: 

> What's @rtega effort in #4091 (comment) all about ? How does it compare to using Mecab ? Is it a Mecab-light ?

Short answer: It does most of what is needed to make lookups work nicely, but it requires you to correctly select the entire word manually. More work would be needed in order to make one-tap lookups work. I believe the code is adapted from Yomichan (a popular GPL'd Japanese pop-up dictionary browser extension).

> Could a small pure Lua solution like this be enough? (a yes/no answer will suffice :)

Yes, on paper it is possible (Yomichan does this) but I expect there may be performance issues because KoReader's dictionary lookups require spawning a new process each time (Yomichan is all in-memory). I will need to look into it a bit further.

<details>
<summary>Explanation why there might be performance issues. tl;dr If possible, using MeCab is preferable.</summary>
The core issue is that it's not clear from just the text whether something is a complete word or not, you need a dictionary to look up parts of the word. For instance while 喋ります is trivial to deinflect, the same word can be written as しゃべります which you need a dictionary to figure out that "しゃべる" is a word and thus り is the stem.
<br>
MeCab is a fairly complicated program which appears to use some kind of complicated algorithm to try to predict how the sentence should be split, so anything we implement is probably going to be slightly more dodgy than that. The most naive implementation would be to try by adding one character at a time, then trying to deinflect it, then doing the lookup -- but this would probably not be performant at all and I expect there are cases where it would fail.
<br>
I expect that Yomichan solves this by using some more clever data structures to store their dictionaries internally, but I'm not sure and will need to look into it. But notably Yomichan does not use StarDict at all (it has its own format that is designed to better support Japanese) so it's possible that StarDict simply cannot be as efficient as Yomichan in this respect. If necessary the plugin could access the dictionary data directly then store the data in a way that we can efficiently look it up, but that's probably really invasive.
</details>

**Comment by rtega** [comment]
> What's @rtega effort in [#4091 (comment)](https://github.com/koreader/koreader/issues/4091#issuecomment-420164656) all about ? How does it compare to using Mecab ? Is it a Mecab-light ? Could a small pure Lua solution like this be enough ? (a yes/no answer will suffice :)
> Or could Mecab be rewritten in Lua ?
> 
> (Sorry, not much brain power available currently to digest the above posts and answer longer :)

The only thing that my code does is deinflect verbs, adverbs and adjectives in Japanese before the word is sent to the dictionary lookup. An equivalent example in English would be: you search for "fought", my code changes that to "fight" and you find the result without fuzzy lookup. For Japanese no such code is available in the dictionary lookup program koreader uses (or any other we could use that I'm aware of). It is very annoying when you have to do that manually since practically every adjective and verb in Japanese is conjugated and doesn't yield a result.

Where mecab comes in is making the selection procedure easier. It breaks up sentences in it's constituent parts. Since Japanese doesn't have spaces, you have must select character by character the exact word you want to look up. Imagine having to that for the previous sentence, this is how it would look in Japanese: "SinceJapanesedoesn'thavespaces,youhavemustselectcharacterbycharactertheexactwordyouwanttolookup." Same for Chinese by the way.

In short: Mecab breaks up sentences, my code deinflects things that are conjugated.

**Comment by cyphar** [comment]
@Frenzie 

> This belongs in the main repo, doesn't it?

I wasn't sure if it being Japanese-specific would discount it from being included in the main repo (@poire-z's response seems to indicate that is the case). If that isn't the case then yes, it would be nice for it to be included in the main repo but I don't mind maintaining it separately.

<details>
<summary>Details about Japanese Grammar and MeCab. tl;dr single-tap dictionary lookups would be fairly easy to implement with MeCab in the Japanese plugin.</summary>
I also just checked the output of MeCab again and actually it does the necessary deinflection for Japanese already (not only that but it's actually easier to just use it for deinflection than it is to use it to pick the right text length to select -- though you could use it for that. But these are details of the plugin that are Japanese-specific and aren't really relevant to anyone other than me.)

For my future reference -- MeCab segments the sentence using the concept of helper verbs (助動詞) which means it actually splits verbs into their stem and then treats the helper verbs as separate words. The solution would simply be to join up all runs of helper verbs after the stem (maybe with special handling for です because that is technically a 助動詞) but as I said this is not really necessary because MeCab also just gives you the dictionary form so we can just use that. In fact this would allow for single-tap lookups to work fairly flawlessly without the need for any deinflection code in the plugin.
</details>

> Is it a problem that it says the last update was from 2013?

No, that's simply the latest version of MeCab. It seems the author has stopped working on it, but the vast majority of Japanese text related projects depend on MeCab (MacOS/iOS use it, as does Google Japanese Input, and many other open source projects like Anki use it as well).

**Comment by rtega** [comment]
> > What's @rtega effort in #4091 (comment) all about ? How does it compare to using Mecab ? Is it a Mecab-light ?
> 
> Short answer: It does most of what is needed to make lookups work nicely, but it requires you to correctly select the entire word manually. More work would be needed in order to make one-tap lookups work. I believe the code is adapted from Yomichan (a popular GPL'd Japanese pop-up dictionary browser extension).

My code was implemented by looking at yomichan.

**Comment by cyphar** [comment]
I looked into it an Yomichan's native parsing appears to do the naive approach I mentioned above, so it all depends on how bad the performance of each dictionary lookup is (if it's not that bad, MeCab would technically not be necessary). I'll give it a shot in the emulator but I wonder how much of a performance hit you'd see on less capable devices.

**Comment by cyphar** [comment]
I looked into it an Yomichan's native parsing appears to do the naive approach I mentioned above, and I've figured out that `sdcv` supports looking up multiple terms at once so with the right tuning it might be possible to avoid the need for MeCab entirely (though on paper it would be nice to be able to use MeCab but maybe that can be figured out in the future if doing it Yomichan-style is workable).
